File: spotify/spotify.ts
#!/usr/bin/env bun
/**
 * Spotify Web API helper — token management + common playlist operations.
 *
 * Run with Bun (no build step): `bun spotify.ts <command>`. Auto-loads .env, so
 * you can just run e.g. `bun spotify.ts me` without exporting anything first.
 *
 * Env vars (from .env, see SKILL.md):
 *   SPOTIFY_CLIENT_ID, SPOTIFY_CLIENT_SECRET   (required)
 *   SPOTIFY_REDIRECT_URI   (default http://127.0.0.1:8888/callback; MUST be registered on the app)
 *   SPOTIFY_PLAYLIST_ID    (optional default playlist for playlist/add-* commands)
 *   SPOTIFY_REFRESH_TOKEN  (optional; with it set the script never needs the browser)
 *   SPOTIFY_SCOPES         (optional; default is the full user-facing Web API scope set)
 *
 * The user token is cached in .spotify_token.json next to this file and refreshed
 * automatically. The FIRST run (no refresh token anywhere) needs interactive
 * authorization: it prints `AUTH_URL <url>` and waits on the local callback server
 * — open that URL in a browser logged into the right Spotify account.
 *
 * Commands:
 *   auth                                          (re)authorize and cache a token
 *   token [--force]                               print a valid access token
 *   me                                            print the current user
 *   playlist [PLAYLIST_ID]                        print playlist name / owner / track total
 *   search "QUERY" [--type album|track] [--limit N]
 *   album-tracks ALBUM                            list a release's tracks
 *   add-album  ALBUM [ALBUM ...] [--playlist ID] [--allow-dupes]
 *   add-tracks TRACK [TRACK ...] [--playlist ID] [--allow-dupes]
 *
 * ALBUM/TRACK accept a raw id, a spotify:album:/spotify:track: URI, or an
 * open.spotify.com/... URL. add-* skip tracks already in the playlist unless
 * --allow-dupes, and append in the given order.
 */
import { existsSync, readFileSync, writeFileSync, chmodSync } from "node:fs";
import { join } from "node:path";

const HERE = import.meta.dir;
const TOKEN_FILE = join(HERE, ".spotify_token.json");
const API = "https://api.spotify.com/v1";
const ACCOUNTS = "https://accounts.spotify.com";

// ---------------------------------------------------------------- env / config
/** Load .env (skill-dir parent or $SKILLS_DIR). Real env vars win over .env. */
function loadDotenv(): void {
  const home = process.env.HOME ?? "";
  const candidates = [
    join(HERE, "..", ".env"),
    join(process.env.SKILLS_DIR ?? join(home, ".claude/skills"), ".env"),
  ];
  for (const path of candidates) {
    if (!existsSync(path)) continue;
    for (const raw of readFileSync(path, "utf8").split("\n")) {
      const line = raw.trim();
      if (!line || line.startsWith("#") || !line.includes("=")) continue;
      const eq = line.indexOf("=");
      const key = line.slice(0, eq).trim();
      let val = line.slice(eq + 1).trim();
      if (/^".*"$/.test(val) || /^'.*'$/.test(val)) val = val.slice(1, -1);
      if (process.env[key] === undefined) process.env[key] = val;
    }
    break;
  }
}
loadDotenv();

function reqEnv(name: string): string {
  const v = process.env[name];
  if (!v) die(`Missing required env var ${name} — copy .env.example to .env and fill it in.`);
  return v;
}
function die(msg: string): never {
  console.error(msg);
  process.exit(1);
}

const CLIENT_ID = reqEnv("SPOTIFY_CLIENT_ID");
const CLIENT_SECRET = reqEnv("SPOTIFY_CLIENT_SECRET");
const REDIRECT_URI = process.env.SPOTIFY_REDIRECT_URI ?? "http://127.0.0.1:8888/callback";
const DEFAULT_PLAYLIST = process.env.SPOTIFY_PLAYLIST_ID;
// Named playlists: every SPOTIFY_PLAYLIST_<NAME> env var (except _ID) becomes the alias
// <name> (lowercased), usable as `--playlist <name>`. SPOTIFY_PLAYLIST_ID is the default
// target and also answers to the alias "main".
const PLAYLIST_ALIASES: Record<string, string> = {};
for (const [k, v] of Object.entries(process.env)) {
  const m = /^SPOTIFY_PLAYLIST_(.+)$/.exec(k);
  if (m && v && m[1].toUpperCase() !== "ID") PLAYLIST_ALIASES[m[1].toLowerCase()] = v;
}
if (DEFAULT_PLAYLIST) PLAYLIST_ALIASES.main ??= DEFAULT_PLAYLIST;
const PLAYLIST_SYNONYMS: Record<string, string> = { running: "sport", run: "sport", todo: "main" };
// Default: full set of user-facing Web API scopes, so a missing scope is never the
// blocker. Editing playlists only strictly needs playlist-modify-*.
const SCOPES =
  process.env.SPOTIFY_SCOPES ??
  [
    "ugc-image-upload", "user-read-playback-state", "user-modify-playback-state",
    "user-read-currently-playing", "app-remote-control", "streaming",
    "playlist-read-private", "playlist-read-collaborative",
    "playlist-modify-private", "playlist-modify-public",
    "user-follow-modify", "user-follow-read", "user-read-playback-position",
    "user-top-read", "user-read-recently-played", "user-library-modify",
    "user-library-read", "user-read-email", "user-read-private",
  ].join(" ");

// ---------------------------------------------------------------- token types
interface RawToken {
  access_token: string;
  token_type?: string;
  expires_in?: number;
  refresh_token?: string;
  scope?: string;
}
interface Token extends RawToken {
  expires_at: number; // unix seconds
}

// ---------------------------------------------------------------- auth / token
function basicAuth(): string {
  return Buffer.from(`${CLIENT_ID}:${CLIENT_SECRET}`).toString("base64");
}

async function postToken(form: Record<string, string>): Promise<RawToken> {
  const res = await fetch(`${ACCOUNTS}/api/token`, {
    method: "POST",
    headers: {
      Authorization: `Basic ${basicAuth()}`,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: new URLSearchParams(form).toString(),
  });
  if (!res.ok) throw new Error(`token endpoint ${res.status}: ${await res.text()}`);
  return (await res.json()) as RawToken;
}

function saveToken(tok: RawToken): Token {
  const out: Token = { ...tok, expires_at: Date.now() / 1000 + (tok.expires_in ?? 3600) - 60 };
  if (!out.refresh_token && existsSync(TOKEN_FILE)) {
    try {
      out.refresh_token = (JSON.parse(readFileSync(TOKEN_FILE, "utf8")) as Token).refresh_token;
    } catch {
      /* ignore unreadable cache */
    }
  }
  writeFileSync(TOKEN_FILE, JSON.stringify(out));
  chmodSync(TOKEN_FILE, 0o600);
  return out;
}

/** Browser consent flow — one-time bootstrap to mint a refresh token. */
function interactiveAuth(): Promise<Token> {
  const { hostname, port } = new URL(REDIRECT_URI);
  const authUrl =
    `${ACCOUNTS}/authorize?` +
    new URLSearchParams({
      client_id: CLIENT_ID,
      response_type: "code",
      redirect_uri: REDIRECT_URI,
      scope: SCOPES,
      show_dialog: "false",
    }).toString();

  return new Promise<Token>((resolve, reject) => {
    const html = (msg: string) =>
      new Response(`<html><body style="font-family:sans-serif;padding:3em"><h2>${msg}</h2></body></html>`, {
        headers: { "Content-Type": "text/html; charset=utf-8" },
      });
    const server = Bun.serve({
      hostname,
      port: Number(port) || 8888,
      async fetch(req) {
        const url = new URL(req.url);
        if (url.pathname !== "/callback") return new Response("not found", { status: 404 });
        const code = url.searchParams.get("code");
        const err = url.searchParams.get("error");
        setTimeout(() => server.stop(true), 250); // let this response flush, then shut down
        if (err || !code) {
          reject(new Error(`authorization failed: ${err ?? "no code returned"}`));
          return html("Authorization failed. You can close this tab.");
        }
        try {
          const tok = saveToken(
            await postToken({ grant_type: "authorization_code", code, redirect_uri: REDIRECT_URI }),
          );
          console.log(`TOKEN_OK scopes=${tok.scope ?? ""}`);
          if (tok.refresh_token) {
            console.log(`REFRESH_TOKEN ${tok.refresh_token}`);
            console.log("-> Save this as SPOTIFY_REFRESH_TOKEN in .env to skip the browser everywhere.");
          }
          resolve(tok);
          return html("Authorized. Return to the terminal.");
        } catch (e) {
          reject(e instanceof Error ? e : new Error(String(e)));
          return html("Token exchange failed. Check the terminal.");
        }
      },
    });
    console.log(`AUTH_URL ${authUrl}`);
    console.log(`Waiting for authorization on ${REDIRECT_URI} ...`);
  });
}

async function refresh(refreshToken: string): Promise<RawToken> {
  const tok = await postToken({ grant_type: "refresh_token", refresh_token: refreshToken });
  tok.refresh_token ??= refreshToken; // Spotify omits it on refresh — keep ours
  return tok;
}

/** valid cached access token -> refresh (env SPOTIFY_REFRESH_TOKEN, else cached) -> browser. */
async function getToken(force = false): Promise<Token> {
  let cached: Token | null = null;
  if (!force && existsSync(TOKEN_FILE)) {
    try {
      cached = JSON.parse(readFileSync(TOKEN_FILE, "utf8")) as Token;
    } catch {
      cached = null;
    }
    if (cached?.access_token && (cached.expires_at ?? 0) > Date.now() / 1000) return cached;
  }
  if (!force) {
    const rt = process.env.SPOTIFY_REFRESH_TOKEN || cached?.refresh_token;
    if (rt) {
      try {
        return saveToken(await refresh(rt));
      } catch (e) {
        console.error(`Token refresh failed; re-authorizing... (${e})`);
      }
    }
  }
  return interactiveAuth();
}

// ---------------------------------------------------------------- api helpers
type Params = Record<string, string | number>;

async function api(method: string, path: string, params?: Params, body?: unknown): Promise<any> {
  const token = (await getToken()).access_token;
  let url = API + path;
  if (params) {
    const qs = new URLSearchParams();
    for (const [k, v] of Object.entries(params)) qs.set(k, String(v));
    url += `?${qs}`;
  }
  while (true) {
    const res = await fetch(url, {
      method,
      headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
      body: body !== undefined ? JSON.stringify(body) : undefined,
    });
    if (res.status === 429) {
      const wait = (Number(res.headers.get("Retry-After") ?? "2") + 1) * 1000;
      console.error(`  rate limited, waiting ${wait}ms`);
      await Bun.sleep(wait);
      continue;
    }
    if (!res.ok) die(`API ${res.status} on ${method} ${path}: ${await res.text()}`);
    const text = await res.text();
    return text ? JSON.parse(text) : {};
  }
}

function spotifyId(s: string, kind: string): string {
  const m = s.trim().match(new RegExp(`(?:spotify:${kind}:|/${kind}/)([A-Za-z0-9]+)`));
  return m ? m[1]! : s.trim();
}

async function paginate(path: string, params: Params, limit: number): Promise<any[]> {
  const out: any[] = [];
  let offset = 0;
  while (true) {
    const page = await api("GET", path, { ...params, limit, offset });
    out.push(...(page.items ?? []));
    if (!page.next) return out;
    offset += limit;
  }
}

/** Resolve a playlist arg: a named alias (sport/classics/main/...), or a raw id/URI/URL. */
function needPlaylist(arg?: string): string {
  const raw = arg ?? DEFAULT_PLAYLIST;
  if (!raw) die("No playlist — pass an id/name/URL or set SPOTIFY_PLAYLIST_ID in .env.");
  const key = raw.toLowerCase();
  const resolved = PLAYLIST_ALIASES[PLAYLIST_SYNONYMS[key] ?? key] ?? raw;
  return spotifyId(resolved, "playlist");
}

async function existingUris(pid: string): Promise<Set<string>> {
  const items = await paginate(`/playlists/${pid}/tracks`, { fields: "items(track(uri)),next" }, 100);
  return new Set(items.filter((it) => it.track?.uri).map((it) => it.track.uri as string));
}

async function addUris(pid: string, uris: string[]): Promise<void> {
  for (let i = 0; i < uris.length; i += 100) {
    await api("POST", `/playlists/${pid}/tracks`, undefined, { uris: uris.slice(i, i + 100) });
  }
}

const pad = (s: unknown, n: number) => String(s).padStart(n);

// ---------------------------------------------------------------- commands
interface Flags {
  playlist?: string;
  type?: string;
  limit?: string;
  "allow-dupes"?: boolean;
  force?: boolean;
}

const commands: Record<string, (pos: string[], flags: Flags) => Promise<void> | void> = {
  auth: async () => {
    await interactiveAuth();
  },

  token: async (_pos, flags) => {
    console.log((await getToken(Boolean(flags.force))).access_token);
  },

  me: async () => {
    const me = await api("GET", "/me");
    console.log(JSON.stringify({ display_name: me.display_name, id: me.id, country: me.country, product: me.product }));
  },

  playlist: async (pos, flags) => {
    const pid = needPlaylist(flags.playlist ?? pos[0]);
    const pl = await api("GET", `/playlists/${pid}`, {
      fields: "name,public,collaborative,owner(display_name,id),tracks(total)",
    });
    console.log(JSON.stringify(pl, null, 2));
  },

  playlists: async () => {
    const entries = Object.entries(PLAYLIST_ALIASES);
    if (!entries.length) return console.log("No named playlists. Add SPOTIFY_PLAYLIST_<NAME> to .env.");
    for (const [name, id] of entries) {
      const pl = await api("GET", `/playlists/${id}`, { fields: "name,tracks(total)" });
      console.log(`${name.padEnd(10)} ${id}  ${pl.name} (${pl.tracks.total})`);
    }
  },

  search: async (pos, flags) => {
    const type = flags.type ?? "album";
    const res = await api("GET", "/search", { q: pos[0] ?? "", type, limit: Number(flags.limit ?? 12) });
    for (const it of res[`${type}s`]?.items ?? []) {
      const arts = (it.artists ?? []).map((a: { name: string }) => a.name).join(", ");
      if (type === "album") {
        console.log(`${it.id} | ${String(it.album_type).padEnd(6)} | tracks=${pad(it.total_tracks, 2)} | ${String(it.release_date ?? "?").padEnd(10)} | ${it.name}  —  [${arts}]`);
      } else {
        console.log(`${it.id} | ${it.name}  —  [${arts}]  (${it.album?.name})`);
      }
    }
  },

  "album-tracks": async (pos) => {
    const aid = spotifyId(pos[0] ?? "", "album");
    for (const t of await paginate(`/albums/${aid}/tracks`, {}, 50)) {
      console.log(`${pad(t.track_number, 2)}. ${t.name}  [${t.uri}]`);
    }
  },

  "add-album": async (pos, flags) => {
    const pid = needPlaylist(flags.playlist);
    const have = flags["allow-dupes"] ? new Set<string>() : await existingUris(pid);
    const toAdd: string[] = [];
    let skipped = 0;
    for (const raw of pos) {
      const aid = spotifyId(raw, "album");
      const tracks = await paginate(`/albums/${aid}/tracks`, {}, 50);
      console.log(`\nAlbum ${aid}: ${tracks.length} tracks`);
      for (const t of tracks) {
        if (have.has(t.uri)) {
          skipped++;
          console.log(`   = already present: ${pad(t.track_number, 2)}. ${t.name}`);
        } else {
          toAdd.push(t.uri);
          have.add(t.uri);
          console.log(`   + ${pad(t.track_number, 2)}. ${t.name}`);
        }
      }
    }
    await addUris(pid, toAdd);
    const pl = await api("GET", `/playlists/${pid}`, { fields: "tracks(total),name" });
    console.log(`\nAdded ${toAdd.length} tracks (${skipped} skipped as already present). '${pl.name}' now has ${pl.tracks.total}.`);
  },

  "add-tracks": async (pos, flags) => {
    const pid = needPlaylist(flags.playlist);
    const have = flags["allow-dupes"] ? new Set<string>() : await existingUris(pid);
    const toAdd: string[] = [];
    let skipped = 0;
    for (const raw of pos) {
      const uri = `spotify:track:${spotifyId(raw, "track")}`;
      if (have.has(uri)) skipped++;
      else {
        toAdd.push(uri);
        have.add(uri);
      }
    }
    await addUris(pid, toAdd);
    const pl = await api("GET", `/playlists/${pid}`, { fields: "tracks(total),name" });
    console.log(`Added ${toAdd.length} tracks (${skipped} skipped). '${pl.name}' now has ${pl.tracks.total}.`);
  },
};

// ---------------------------------------------------------------- arg parsing
const BOOL_FLAGS = new Set(["allow-dupes", "force"]);

function parseArgs(args: string[]): { positionals: string[]; flags: Flags } {
  const positionals: string[] = [];
  const flags: Flags = {};
  for (let i = 0; i < args.length; i++) {
    const a = args[i]!;
    if (a.startsWith("--")) {
      const name = a.slice(2);
      if (BOOL_FLAGS.has(name)) (flags as Record<string, unknown>)[name] = true;
      else (flags as Record<string, unknown>)[name] = args[++i];
    } else {
      positionals.push(a);
    }
  }
  return { positionals, flags };
}

const [cmd, ...rest] = Bun.argv.slice(2);
const handler = cmd ? commands[cmd] : undefined;
if (!handler) {
  die(`Unknown command: ${cmd ?? "(none)"}\nCommands: ${Object.keys(commands).join(", ")}`);
}
const { positionals, flags } = parseArgs(rest);
await handler(positionals, flags);

File: spotify/spotify.py
#!/usr/bin/env python3
"""Spotify Web API helper — token management + common playlist operations.

Pure Python 3 standard library — no pip installs, no build step, runs anywhere
python3 exists (macOS + the cloud sandbox both ship it):

    python3 spotify.py <command>

Auto-loads .env, so you can run it without exporting anything first.

Env vars (from .env, see .env.example):
  SPOTIFY_CLIENT_ID, SPOTIFY_CLIENT_SECRET   (required)
  SPOTIFY_REDIRECT_URI    (default http://127.0.0.1:8888/callback; MUST be registered on the app)
  SPOTIFY_PLAYLIST_ID     (default playlist for playlist/add-* commands)
  SPOTIFY_PLAYLIST_<NAME> (named playlists, each usable as --playlist <name>)
  SPOTIFY_REFRESH_TOKEN   (optional; with it set the script never needs the browser)
  SPOTIFY_SCOPES          (optional; default is the full user-facing Web API scope set)

The user token is cached in .spotify_token.json next to this file and refreshed
automatically. The FIRST run (no refresh token anywhere) needs interactive
authorization: it prints `AUTH_URL <url>` and waits on the local callback server —
open that URL in a browser logged into the right Spotify account.

Commands:
  auth                                          (re)authorize and cache a token
  token [--force]                               print a valid access token
  me                                            print the current user
  playlist [ID|NAME]                            print playlist name / owner / track total
  playlists                                     list named playlists
  search "QUERY" [--type album|track] [--limit N]
  album-tracks ALBUM                            list a release's tracks
  add-album  ALBUM [ALBUM ...] [--playlist ID|NAME] [--allow-dupes]
  add-tracks TRACK [TRACK ...] [--playlist ID|NAME] [--allow-dupes]
  remove-tracks TRACK [TRACK ...] [--playlist ID|NAME]
  move TRACK [TRACK ...] --from ID|NAME --to ID|NAME [--allow-dupes]

ALBUM/TRACK accept a raw id, a spotify:album:/spotify:track: URI, or an
open.spotify.com/... URL.
"""
import argparse
import base64
import http.server
import json
import os
import re
import sys
import threading
import time
import urllib.error
import urllib.parse
import urllib.request

HERE = os.path.dirname(os.path.abspath(__file__))
TOKEN_FILE = os.path.join(HERE, ".spotify_token.json")
API = "https://api.spotify.com/v1"
ACCOUNTS = "https://accounts.spotify.com"


def die(msg):
    print(msg, file=sys.stderr)
    sys.exit(1)


def load_dotenv():
    """Load .env (skill-dir parent or $SKILLS_DIR). Real env vars win over .env."""
    candidates = [
        os.path.join(os.path.dirname(HERE), ".env"),
        os.path.join(os.environ.get("SKILLS_DIR", os.path.expanduser("~/.claude/skills")), ".env"),
    ]
    for path in candidates:
        if not os.path.exists(path):
            continue
        with open(path) as f:
            for raw in f:
                line = raw.strip()
                if not line or line.startswith("#") or "=" not in line:
                    continue
                key, val = line.split("=", 1)
                key, val = key.strip(), val.strip()
                if (val[:1], val[-1:]) in (('"', '"'), ("'", "'")):
                    val = val[1:-1]
                os.environ.setdefault(key, val)
        break


def env(name, default=None, required=False):
    v = os.environ.get(name, default)
    if required and not v:
        die(f"Missing required env var {name} — copy .env.example to .env and fill it in.")
    return v


load_dotenv()
CLIENT_ID = env("SPOTIFY_CLIENT_ID", required=True)
CLIENT_SECRET = env("SPOTIFY_CLIENT_SECRET", required=True)
REDIRECT_URI = env("SPOTIFY_REDIRECT_URI", "http://127.0.0.1:8888/callback")
DEFAULT_PLAYLIST = env("SPOTIFY_PLAYLIST_ID")
# Default: full set of user-facing Web API scopes, so a missing scope is never the
# blocker. Editing playlists only strictly needs playlist-modify-*.
SCOPES = env("SPOTIFY_SCOPES",
             "ugc-image-upload user-read-playback-state user-modify-playback-state "
             "user-read-currently-playing app-remote-control streaming "
             "playlist-read-private playlist-read-collaborative "
             "playlist-modify-private playlist-modify-public "
             "user-follow-modify user-follow-read user-read-playback-position "
             "user-top-read user-read-recently-played user-library-modify "
             "user-library-read user-read-email user-read-private")

# Named playlists: every SPOTIFY_PLAYLIST_<NAME> env var (except _ID) becomes the alias
# <name> (lowercased), usable as `--playlist <name>`. SPOTIFY_PLAYLIST_ID is the default
# target and also answers to the alias "main".
PLAYLIST_ALIASES = {}
for _k, _v in os.environ.items():
    _m = re.match(r"^SPOTIFY_PLAYLIST_(.+)$", _k)
    if _m and _v and _m.group(1).upper() != "ID":
        PLAYLIST_ALIASES[_m.group(1).lower()] = _v
if DEFAULT_PLAYLIST:
    PLAYLIST_ALIASES.setdefault("main", DEFAULT_PLAYLIST)
PLAYLIST_SYNONYMS = {"running": "sport", "run": "sport", "todo": "main"}


# --- auth / token ---
def _basic():
    return base64.b64encode(f"{CLIENT_ID}:{CLIENT_SECRET}".encode()).decode()


def _post_token(form):
    req = urllib.request.Request(
        ACCOUNTS + "/api/token", data=urllib.parse.urlencode(form).encode(),
        headers={"Authorization": "Basic " + _basic(),
                 "Content-Type": "application/x-www-form-urlencoded"})
    with urllib.request.urlopen(req) as r:
        return json.load(r)


def save_token(tok):
    out = dict(tok)
    out["expires_at"] = time.time() + out.get("expires_in", 3600) - 60
    if not out.get("refresh_token") and os.path.exists(TOKEN_FILE):
        try:  # refresh responses omit the refresh_token — keep the existing one
            out["refresh_token"] = json.load(open(TOKEN_FILE)).get("refresh_token")
        except Exception:
            pass
    with open(TOKEN_FILE, "w") as f:
        json.dump(out, f)
    os.chmod(TOKEN_FILE, 0o600)  # restrict the refresh-token file to the owner
    return out


def interactive_auth():
    u = urllib.parse.urlparse(REDIRECT_URI)
    host, port = (u.hostname or "127.0.0.1"), (u.port or 8888)
    auth_url = ACCOUNTS + "/authorize?" + urllib.parse.urlencode({
        "client_id": CLIENT_ID, "response_type": "code",
        "redirect_uri": REDIRECT_URI, "scope": SCOPES, "show_dialog": "false"})
    got = {}

    class Handler(http.server.BaseHTTPRequestHandler):
        def log_message(self, *a):
            pass

        def _html(self, msg):
            self.send_response(200)
            self.send_header("Content-Type", "text/html; charset=utf-8")
            self.end_headers()
            self.wfile.write(("<html><body style='font-family:sans-serif;padding:3em'>"
                              f"<h2>{msg}</h2></body></html>").encode())

        def do_GET(self):
            q = urllib.parse.parse_qs(urllib.parse.urlparse(self.path).query)
            if "code" in q:
                got["code"] = q["code"][0]
                self._html("Authorized. Return to the terminal.")
            else:
                got["error"] = q.get("error", ["unknown"])[0]
                self._html("Authorization failed. You can close this tab.")
            threading.Thread(target=self.server.shutdown).start()

    srv = http.server.HTTPServer((host, port), Handler)
    print("AUTH_URL " + auth_url, flush=True)
    print(f"Waiting for authorization on {REDIRECT_URI} ...", flush=True)
    srv.serve_forever()
    if "code" not in got:
        die("Authorization failed: " + got.get("error", "no code returned"))
    tok = save_token(_post_token({"grant_type": "authorization_code",
                                  "code": got["code"], "redirect_uri": REDIRECT_URI}))
    print("TOKEN_OK scopes=" + tok.get("scope", ""), flush=True)
    if tok.get("refresh_token"):
        print("REFRESH_TOKEN " + tok["refresh_token"], flush=True)
        print("-> Save this as SPOTIFY_REFRESH_TOKEN in .env to skip the browser everywhere.", flush=True)
    return tok


def _refresh(refresh_token):
    tok = _post_token({"grant_type": "refresh_token", "refresh_token": refresh_token})
    tok.setdefault("refresh_token", refresh_token)
    return tok


def get_token(force=False):
    """valid cached access token -> refresh (env SPOTIFY_REFRESH_TOKEN, else cached) -> browser."""
    cached = None
    if not force and os.path.exists(TOKEN_FILE):
        try:
            cached = json.load(open(TOKEN_FILE))
        except Exception:
            cached = None
        if cached and cached.get("access_token") and cached.get("expires_at", 0) > time.time():
            return cached
    if not force:
        rt = os.environ.get("SPOTIFY_REFRESH_TOKEN") or (cached.get("refresh_token") if cached else None)
        if rt:
            try:
                return save_token(_refresh(rt))
            except urllib.error.HTTPError as e:
                print(f"Token refresh failed ({e.code}); re-authorizing...", file=sys.stderr)
    return interactive_auth()


# --- api ---
def api(method, path, params=None, body=None):
    token = get_token()["access_token"]
    url = API + path + ("?" + urllib.parse.urlencode(params) if params else "")
    data = json.dumps(body).encode() if body is not None else None
    req = urllib.request.Request(url, data=data, method=method,
                                 headers={"Authorization": "Bearer " + token,
                                          "Content-Type": "application/json"})
    while True:
        try:
            with urllib.request.urlopen(req) as r:
                txt = r.read().decode()
                return json.loads(txt) if txt else {}
        except urllib.error.HTTPError as e:
            if e.code == 429:
                wait = int(e.headers.get("Retry-After", "2")) + 1
                print(f"  rate limited, waiting {wait}s", file=sys.stderr)
                time.sleep(wait)
                continue
            die(f"API {e.code} on {method} {path}: {e.read().decode()}")


def spotify_id(s, kind):
    m = re.search(r"(?:spotify:%s:|/%s/)([A-Za-z0-9]+)" % (kind, kind), s.strip())
    return m.group(1) if m else s.strip()


def paginate(path, params, limit):
    out, off, params = [], 0, dict(params)
    params["limit"] = limit
    while True:
        params["offset"] = off
        page = api("GET", path, params)
        out.extend(page.get("items", []))
        if not page.get("next"):
            return out
        off += limit


def need_playlist(arg):
    """Resolve a playlist arg: a named alias (sport/classics/main/...), or a raw id/URI/URL."""
    raw = arg or DEFAULT_PLAYLIST
    if not raw:
        die("No playlist — pass an id/name/URL or set SPOTIFY_PLAYLIST_ID in .env.")
    key = raw.lower()
    resolved = PLAYLIST_ALIASES.get(PLAYLIST_SYNONYMS.get(key, key), raw)
    return spotify_id(resolved, "playlist")


def existing_uris(pid):
    items = paginate(f"/playlists/{pid}/tracks", {"fields": "items(track(uri)),next"}, 100)
    return {it["track"]["uri"] for it in items if it.get("track") and it["track"].get("uri")}


def add_uris(pid, uris):
    for i in range(0, len(uris), 100):
        api("POST", f"/playlists/{pid}/tracks", body={"uris": uris[i:i + 100]})


def remove_uris(pid, uris):
    for i in range(0, len(uris), 100):
        api("DELETE", f"/playlists/{pid}/tracks",
            body={"tracks": [{"uri": u} for u in uris[i:i + 100]]})


# --- commands ---
def cmd_auth(a):
    interactive_auth()


def cmd_token(a):
    print(get_token(force=a.force)["access_token"])


def cmd_me(a):
    me = api("GET", "/me")
    print(json.dumps({k: me.get(k) for k in ("display_name", "id", "country", "product")},
                     ensure_ascii=False))


def cmd_playlist(a):
    pid = need_playlist(a.playlist)
    pl = api("GET", f"/playlists/{pid}",
             {"fields": "name,public,collaborative,owner(display_name,id),tracks(total)"})
    print(json.dumps(pl, ensure_ascii=False, indent=2))


def cmd_playlists(a):
    if not PLAYLIST_ALIASES:
        print("No named playlists. Add SPOTIFY_PLAYLIST_<NAME> to .env.")
        return
    for name, pid in PLAYLIST_ALIASES.items():
        pl = api("GET", f"/playlists/{pid}", {"fields": "name,tracks(total)"})
        print(f"{name:10} {pid}  {pl['name']} ({pl['tracks']['total']})")


def cmd_search(a):
    res = api("GET", "/search", {"q": a.query, "type": a.type, "limit": a.limit})
    for it in res.get(a.type + "s", {}).get("items", []):
        arts = ", ".join(x["name"] for x in it.get("artists", []))
        if a.type == "album":
            print(f"{it['id']} | {it['album_type']:6} | tracks={it['total_tracks']:2} | "
                  f"{it.get('release_date', '?'):10} | {it['name']}  —  [{arts}]")
        else:
            print(f"{it['id']} | {it['name']}  —  [{arts}]  ({it['album']['name']})")


def cmd_album_tracks(a):
    for t in paginate(f"/albums/{spotify_id(a.album, 'album')}/tracks", {}, 50):
        print(f"{t['track_number']:2}. {t['name']}  [{t['uri']}]")


def cmd_add_album(a):
    pid = need_playlist(a.playlist)
    have = set() if a.allow_dupes else existing_uris(pid)
    to_add, skipped = [], 0
    for raw in a.album:
        aid = spotify_id(raw, "album")
        tracks = paginate(f"/albums/{aid}/tracks", {}, 50)
        print(f"\nAlbum {aid}: {len(tracks)} tracks")
        for t in tracks:
            if t["uri"] in have:
                skipped += 1
                print(f"   = already present: {t['track_number']:2}. {t['name']}")
            else:
                to_add.append(t["uri"])
                have.add(t["uri"])
                print(f"   + {t['track_number']:2}. {t['name']}")
    add_uris(pid, to_add)
    pl = api("GET", f"/playlists/{pid}", {"fields": "tracks(total),name"})
    print(f"\nAdded {len(to_add)} tracks ({skipped} skipped as already present). "
          f"'{pl['name']}' now has {pl['tracks']['total']}.")


def cmd_add_tracks(a):
    pid = need_playlist(a.playlist)
    have = set() if a.allow_dupes else existing_uris(pid)
    to_add, skipped = [], 0
    for raw in a.track:
        uri = "spotify:track:" + spotify_id(raw, "track")
        if uri in have:
            skipped += 1
        else:
            to_add.append(uri)
            have.add(uri)
    add_uris(pid, to_add)
    pl = api("GET", f"/playlists/{pid}", {"fields": "tracks(total),name"})
    print(f"Added {len(to_add)} tracks ({skipped} skipped). '{pl['name']}' now has {pl['tracks']['total']}.")


def cmd_remove_tracks(a):
    pid = need_playlist(a.playlist)
    uris = ["spotify:track:" + spotify_id(raw, "track") for raw in a.track]
    remove_uris(pid, uris)
    pl = api("GET", f"/playlists/{pid}", {"fields": "tracks(total),name"})
    print(f"Removed {len(uris)} track(s). '{pl['name']}' now has {pl['tracks']['total']}.")


def cmd_move(a):
    src = need_playlist(a.src)
    dst = need_playlist(a.dst)
    uris = ["spotify:track:" + spotify_id(raw, "track") for raw in a.track]
    have = set() if a.allow_dupes else existing_uris(dst)
    to_add = [u for u in uris if u not in have]
    add_uris(dst, to_add)       # destination first, so nothing is lost if it fails
    remove_uris(src, uris)
    s = api("GET", f"/playlists/{src}", {"fields": "tracks(total),name"})
    d = api("GET", f"/playlists/{dst}", {"fields": "tracks(total),name"})
    print(f"Moved {len(uris)} track(s): '{s['name']}' ({s['tracks']['total']}) -> "
          f"'{d['name']}' ({d['tracks']['total']}); {len(to_add)} added, "
          f"{len(uris) - len(to_add)} already in dest.")


def main():
    p = argparse.ArgumentParser(description="Spotify Web API helper")
    sub = p.add_subparsers(dest="cmd", required=True)
    sub.add_parser("auth").set_defaults(fn=cmd_auth)
    sp = sub.add_parser("token"); sp.add_argument("--force", action="store_true"); sp.set_defaults(fn=cmd_token)
    sub.add_parser("me").set_defaults(fn=cmd_me)
    sp = sub.add_parser("playlist"); sp.add_argument("playlist", nargs="?"); sp.set_defaults(fn=cmd_playlist)
    sub.add_parser("playlists").set_defaults(fn=cmd_playlists)
    sp = sub.add_parser("search"); sp.add_argument("query")
    sp.add_argument("--type", default="album", choices=["album", "track"])
    sp.add_argument("--limit", type=int, default=12); sp.set_defaults(fn=cmd_search)
    sp = sub.add_parser("album-tracks"); sp.add_argument("album"); sp.set_defaults(fn=cmd_album_tracks)
    sp = sub.add_parser("add-album"); sp.add_argument("album", nargs="+")
    sp.add_argument("--playlist"); sp.add_argument("--allow-dupes", action="store_true")
    sp.set_defaults(fn=cmd_add_album)
    sp = sub.add_parser("add-tracks"); sp.add_argument("track", nargs="+")
    sp.add_argument("--playlist"); sp.add_argument("--allow-dupes", action="store_true")
    sp.set_defaults(fn=cmd_add_tracks)
    sp = sub.add_parser("remove-tracks"); sp.add_argument("track", nargs="+")
    sp.add_argument("--playlist"); sp.set_defaults(fn=cmd_remove_tracks)
    sp = sub.add_parser("move"); sp.add_argument("track", nargs="+")
    sp.add_argument("--from", dest="src", required=True)
    sp.add_argument("--to", dest="dst", required=True)
    sp.add_argument("--allow-dupes", action="store_true"); sp.set_defaults(fn=cmd_move)
    a = p.parse_args()
    a.fn(a)


if __name__ == "__main__":
    main()
